README.md
# Akka Management Cluster HTTP

Akka.NET Management Cluster HTTP, a port of the popular Java/Scala [library](https://github.com/akka/akka-management) to .NET, is a management extension that allows you interaction with an `akka-cluster` through an HTTP interface. This management extension exposes different operations to manage nodes in a cluster.

The operations exposed comparable to the JMX interface `akka-cluster` provides.

![alt text](doc/screenshot.png)

## API Definition

The following table describes the usage of the API:

| Path | HTTP method | Required form fields | Description 
| ------------- | ------------- | ------------- | ------------- 
| `/cluster/members/`  | GET | None | Returns the status of the Cluster in JSON format.
| `/cluster/members/` | GET | address: `{address}` | Returns the status of `{address}` in the Cluster
| `/cluster/members/`  | POST | address: `{address}` | Executes join operation in cluster for the provided `{address}`. 
| `/cluster/members/`  | PUT | address: `{address}`, operation: `down` | Executes down operation in cluster for provided `{address}`. 
| `/cluster/members/`  | PUT | address: `{address}`, operation: `leave` | Executes leave operation in cluster for provided `{address}`. 
| `/cluster/shards/{name}` | GET | None | Returns shard info for the shard region with the provided `{name}`

The expected format of address follows the Cluster URI convention. Example: `akka://Main@myhostname.com:3311`

#### GET /cluster/members responses

| Response code | Description
| ------------- | -----------
| 200           | Status of cluster in JSON format
| 500           | Something went wrong. Cluster might be shutdown.

Example response:

```
{
  "selfNode": "akka.tcp://test@10.10.10.10:1111",
  "members": [
    {
      "node": "akka.tcp://test@10.10.10.10:1111",
      "nodeUid": "1116964444",
      "status": "Up",
      "roles": []
    }
  ],
  "unreachable": [],
  "leader: "akka.tcp://test@10.10.10.10:1111",
  "oldest: "akka.tcp://test@10.10.10.10:1111"
}
```

#### GET /cluster/members/?address=\{address\} responses

| Response code | Description
| ------------- | -----------
| 200           | Status of cluster in JSON format
| 404           | No member was found in the cluster for the given `{address}`.
| 500           | Something went wrong. Cluster might be shutdown.

Example response:

```
{
  "node": "akka.tcp://test@10.10.10.10:1111",
  "nodeUid": "-169203556",
  "status": "Up",
  "roles": []
}
```

#### POST /cluster/members/ responses

| Response code | Description
| ------------- | -----------
| 200           | Executing join operation.
| 500           | Something went wrong. Cluster might be shutdown.

Example response:

```
Joining akka.tcp://test@10.10.10.10:111
```

#### PUT /cluster/members/ responses

| Response code | Operation | Description
| ------------- | --------- | -----------
| 200           | Down      | Executing down operation.
| 200           | Leave     | Executing leave operation.
| 400           |           | Operation supplied in `operation` form field is not supported.
| 404           |           | No member was found in the cluster for the given `{address}`
| 500           |           | Something went wrong. Cluster might be shutdown.

Example response:

```
Downing akka.tcp://test@10.10.10.10:111
```


#### GET /cluster/shards/\{name\} responses

| Response code | Description
| ------------- | -----------
| 200           | Shard region information in JSON format
| 404           | No shard region was found on the node for the given `{name}`

Example response:

```
{
  "regions": [
    {
      "shardId": "1234",
      "numEntities": 30
    }
  ]
}
```

RELEASE_NOTES.md
#### 0.8.1 January 26 2023 ####

Upgraded to Akka v1.4.48

#### 0.8.0 July 16 2021 ####

Upgraded to Akka v1.4.21
Support .NET Core 3.1

#### 0.7.4 May 6 2021 ####

Upgraded to Akka v1.4.19

#### 0.7.3 November 28 2020 ####

Upgraded to Akka v1.4.12

#### 0.7.2 July 25 2020 ####

NetStandard 2.0

#### 0.7.1 May 5 2020 ####

Upgraded to Akka v1.3.18

#### 0.7.0 April 27 2019 ####

Upgraded to Akka v1.3.12

src/Akka.Cluster.Management/Routes.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using Akka.Actor;
using Akka.Cluster.Sharding;
using Akka.Http.Dsl.Model;
using Akka.IO;

namespace Akka.Cluster.Management
{
    public class Routes
    {
        private readonly Cluster cluster;
        //private readonly IActorRef routeActor;

        private static readonly HttpResponse NotFound = HttpResponse.Create()
            .WithStatus(404)
            .WithEntity("Unknown resource!");

        /// <summary>
        /// Creates an instance of [[ClusterHttpManagementRoutes]] to manage the specified
        /// [[akka.cluster.Cluster]] instance. This version does not provide Basic Authentication.
        /// </summary>
        public static Routes Create(Cluster cluster) => new(cluster);

        /// <summary>
        /// Creates an instance of [[ClusterHttpManagementRoutes]] to manage the specified
        /// [[akka.cluster.Cluster]] instance. This version does not provide Basic Authentication.
        /// </summary>
        // private Routes(ActorSystem system) =>
        //     routeActor = system.ActorOf(ClusterHttpManagementRoutes.Props(Cluster.Get(system)), "routes");

        private Routes(Cluster cluster) => this.cluster = cluster;

        public Task<HttpResponse> RequestHandler(HttpRequest request) => request.Path switch
        {
            "/members" => Task.FromResult(NotFound),
            "/shards/{name}" => RouteGetShardInfo(""),
            _ => Task.FromResult(NotFound)
        };

        private async Task<HttpResponse> RouteGetShardInfo(string shardRegionName)
        {
            try
            {
                var stats = await ClusterSharding.Get(cluster.System)
                    .ShardRegion(shardRegionName)
                    .Ask<ShardRegionStats>(GetShardRegionStats.Instance);

                var entity = new ShardDetails(stats.Stats.Select(stat =>
                    new ShardRegionInfo(stat.Key, stat.Value)).ToArray());

                // TODO
                return HttpResponse.Create(200, HttpEntity.Create(ByteString.FromBytes(entity)));
            }
            catch (AskTimeoutException)
            {
                // var entity = new ClusterHttpManagementMessage($"Shard Region [{shardRegionName}] not responding, may have been terminated");

                return HttpResponse.Create(404, new ResponseEntity(
                    "text/plain(UTF-8)",
                    ByteString.FromString($"Shard Region [{shardRegionName}] not responding, may have been terminated")));
            }
            catch (Exception)
            {
                // var entity = new ClusterHttpManagementMessage($"Shard Region [{shardRegionName}] is not started");

                return HttpResponse.Create(404, new ResponseEntity(
                    "text/plain(UTF-8)",
                    ByteString.FromString($"Shard Region [{shardRegionName}] is not started")));
            }
        }

        // private async Task<HttpResponse> HandleGetShardInfo(string name)
        // {
        //     try
        //     {
        //         var response = await routeActor.Ask<Complete>(new GetShardInfo(name), TimeSpan.FromSeconds(3));
        //         return response switch
        //         {
        //             Complete.Success => HttpResponse.Create(), // TODO
        //             Complete.Failure failure => HttpResponse.Create(404, new ResponseEntity("text/plain(UTF-8)", ByteString.FromString(failure.Reason))),
        //             _ => throw new InvalidOperationException("Something went wrong. Cluster might be shutdown.")
        //         };
        //     }
        //     catch
        //     {
        //         return HttpResponse.Create().WithStatus(500);
        //     }
        // }

        private static ClusterMember MemberToClusterMember(Member member) =>
            new(member.UniqueAddress.Address.ToString(), member.UniqueAddress.Uid.ToString(), member.Status.ToString(), member.Roles.ToArray());
    }
}